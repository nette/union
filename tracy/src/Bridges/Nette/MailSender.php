<?php

/**
 * This file is part of the Tracy (https://tracy.nette.org)
 * Copyright (c) 2004 David Grudl (https://davidgrudl.com)
 */

namespace Tracy\Bridges\Nette;

use Nette;
use Tracy\Dumper;
use Tracy\Helpers;


/**
 * Tracy logger bridge for Nette Mail.
 */
class MailSender
{
	use Nette\SmartObject;

	/** @var Nette\Mail\IMailer */
	private $mailer;

	/** @var string|null sender of email notifications */
	private $fromEmail;


	public function __construct(Nette\Mail\IMailer $mailer, $fromEmail = null)
	{
		$this->mailer = $mailer;
		$this->fromEmail = $fromEmail;
	}


	/**
	 * @param  string|\Exception|\Throwable  $message
	 * @param  string  $email
	 * @return void
	 */
	public function send($message, $email)
	{
		$host = preg_replace('#[^\w.-]+#', '', isset($_SERVER['HTTP_HOST']) ? $_SERVER['HTTP_HOST'] : php_uname('n'));

		$mail = new Nette\Mail\Message;
		$mail->setHeader('X-Mailer', 'Tracy');
		$mail->setFrom($this->fromEmail ?: "noreply@$host");
		$mail->addTo($email);
		$mail->setSubject('PHP: An error occurred on the server ' . $host);
		$mail->setBody(static::formatMessage($message) . "\n\nsource: " . Helpers::getSource());

		$this->mailer->send($mail);
	}


	/**
	 * @param  mixed  $message
	 * @return string
	 */
	private static function formatMessage($message)
	{
		if ($message instanceof \Exception || $message instanceof \Throwable) {
			while ($message) {
				$tmp[] = ($message instanceof \ErrorException
					? Helpers::errorTypeToString($message->getSeverity()) . ': ' . $message->getMessage()
					: Helpers::getClass($message) . ': ' . $message->getMessage() . ($message->getCode() ? ' #' . $message->getCode() : '')
				) . ' in ' . $message->getFile() . ':' . $message->getLine();
				$message = $message->getPrevious();
			}
			$message = implode("\ncaused by ", $tmp);

		} elseif (!is_string($message)) {
			$message = Dumper::toText($message);
		}

		return trim($message);
	}
}
