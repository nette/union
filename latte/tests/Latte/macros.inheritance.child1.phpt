<?php

/**
 * Test: Nette\Latte\Engine: {extends ...} test I.
 *
 * @author     David Grudl
 */

use Nette\Latte,
	Tester\Assert;


require __DIR__ . '/../bootstrap.php';


$latte = new Latte\Engine;
$latte->setTempDirectory(TEMP_DIR);

Assert::matchFile(
	__DIR__ . '/expected/macros.inheritance.child1.child.phtml',
	$latte->compile(__DIR__ . '/templates/inheritance.child1.latte')
);
Assert::matchFile(
	__DIR__ . '/expected/macros.inheritance.child1.html',
	$latte->renderToString(
		__DIR__ . '/templates/inheritance.child1.latte',
		array('people' => array('John', 'Mary', 'Paul'))
	)
);
Assert::matchFile(
	__DIR__ . '/expected/macros.inheritance.child1.parent.phtml',
	file_get_contents($latte->getCacheFile(__DIR__ . '/templates/inheritance.parent.latte'))
);
