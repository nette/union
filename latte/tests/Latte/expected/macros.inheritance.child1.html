<!DOCTYPE html>
<head>
	<title>Homepage | My websiteMy website</title>
</head>

<body>
	<div id="sidebar">
		<ul>
			<li><a href="/">Homepage</a></li>
		</ul>
	</div>

	<div id="content">
	<ul>
		<li>John</li>
		<li>Mary</li>
		<li>Paul</li>
	</ul>
	Parent: NULL
	</div>
</body>
</html>
Parent: main/extends
